tls/s2n_x509_validator.c

```diff
@@ -436,26 +436,59 @@ S2N_RESULT s2n_x509_validator_check_cert_preferences(struct s2n_connection *conn
      *# "signature_algorithms" extension also applies to signatures appearing in
      *# certificates.
      */
-    if (security_policy->certificate_signature_preferences == NULL) {
+    if (security_policy->certificate_signature_preferences == NULL
+            && security_policy->certificate_key_preferences == NULL) {
         return S2N_RESULT_OK;
     }
 
     struct s2n_cert_info info = { 0 };
     RESULT_GUARD(s2n_openssl_x509_get_cert_info(cert, &info));
 
-    if (info.self_signed) {
-        return S2N_RESULT_OK;
-    }
-
     /* Ensure that the certificate signature does not use SHA-1. While this check
      * would ideally apply to all connections, we only enforce it when certificate
      * preferences exist to stay backwards compatible.
      */
-    if (conn->actual_protocol_version == S2N_TLS13) {
+    if (!info.self_signed && conn->actual_protocol_version == S2N_TLS13) {
         RESULT_ENSURE(info.signature_digest_nid != NID_sha1, S2N_ERR_CERT_UNTRUSTED);
     }
 
-    RESULT_ENSURE_OK(s2n_security_policy_validate_cert_signature(security_policy, &info), S2N_ERR_CERT_UNTRUSTED);
+    if (!info.self_signed) {
+        RESULT_ENSURE_OK(s2n_security_policy_validate_cert_signature(security_policy, &info), S2N_ERR_CERT_UNTRUSTED);
+    }
+    RESULT_ENSURE_OK(s2n_security_policy_validate_cert_key(security_policy, &info), S2N_ERR_CERT_UNTRUSTED);
+
+    return S2N_RESULT_OK;
+}
+
+/* Validates that the root certificate uses a key allowed by the security policy
+ * certificate preferences.
+ */
+static S2N_RESULT s2n_x509_validator_check_root_cert(struct s2n_x509_validator *validator, struct s2n_connection *conn)
+{
+    RESULT_ENSURE_REF(validator);
+    RESULT_ENSURE_REF(conn);
+
+    const struct s2n_security_policy *sp = NULL;
+    RESULT_GUARD_POSIX(s2n_connection_get_security_policy(conn, &sp));
```

Comment:
Nit: paranoia. Seg faults are bad.

Also, let's not get into the habit of tiny, unreadable variable names. If you saw "sp" further down the method, you wouldn't guess "security policy". It gets particularly bad if you have more than one two-letter name. Use something like "security_policy", or "policy" if you're trying to be brief.
```suggestion
    const struct s2n_security_policy *policy = NULL;
    RESULT_GUARD_POSIX(s2n_connection_get_security_policy(conn, &sp));
    RESULT_ENSURE_REF(sp);
```